### DISCUSSION TITLE
cache in graphqlite

### DISCUSSION THREAD

**Discussion by dhiabenjaber** [thecodingmachine/graphqlite#465] [Q&A]
i want to understand how to manage cache with graphiqlte (other no framework) and if there is an example of code cache with memcached, redis ... and what is the purpose of the Adapter

**Comment by oojacoboo** [comment]
You can read the docs on PSR-16 cache integration:
https://graphqlite.thecodingmachine.io/docs/next/other-frameworks

All the cache is handled internally, so you just need to pass a PSR-16 compatible cache adapter.  Cache doesn't need to persist across servers, so Redis is a bad choice.  Stick with something local, like APCU.

**Reply by andrew-demb** [reply]
@oojacoboo you mean cache for schema types, but there is another place to use cache - for execution results.

Probably @dhiabenjaber requesting examples about result cache implementation (which is pretty common to have a cross-server cache server).

**Comment by dhiabenjaber** [comment]
@andrew-demb yes i wan a implement cache in real world application (e-commerce website) to improve performance of the request when i execute a request time to respond is very high

**Comment by oojacoboo** [comment]
@dhiabenjaber what have you implemented so far?  Are you using opcache?  What about pre-loading?  Are you spinning up a lot of instances?  Is this every request?  Have you checked your database cache layers?

Please see XDebug's profiler:
https://xdebug.org/docs/profiler

Can you please get us a summary of these results on a troublesome request?

**Comment by dhiabenjaber** [comment]
@oojacoboo this is screenshots from xdebug profiler 
![image](https://user-images.githubusercontent.com/74189546/168123879-cc60a212-45c6-4758-891a-b11740ae5b94.png)
![Capturexb](https://user-images.githubusercontent.com/74189546/168123958-23ecb0b6-e3b6-4704-861a-f41f5aad48fd.PNG)
![image](https://user-images.githubusercontent.com/74189546/168124165-73be9808-ba29-4507-96ba-e561c2466555.png)
![image](https://user-images.githubusercontent.com/74189546/168124263-a5d1abdf-1de0-4f8c-880e-2faed01b4e2e.png)
i do know the problem is from the psr16cache from symfony because  the framework is not symfony (implementation of graphqlite with no any framework) if you have any idea please help because the time of response is too much .Or if we can do a call to explain to you the problem And Thanks anyway ;)

**Comment by oojacoboo** [comment]
@dhiabenjaber what does your GraphQLite configuration look like?  What cache implementation are you using here?

**Comment by dhiabenjaber** [comment]
![image](https://user-images.githubusercontent.com/74189546/168126152-23026391-5f22-427b-8842-31578fbf83bc.png)
i wan change the ' $cache=new Psr16Cache(new ArrayAdapter());' with an other thing because i don't work with symfony

**Comment by oojacoboo** [comment]
Please try `Symfony\Component\Cache\Adapter\ApcuAdapter` instead of `ArrayAdapter`

**Comment by oojacoboo** [comment]
Also, that's just a Symfony component.  You don't need to be using the Symfnoy framework to use Symfony component libs.  They're standalone libs.

**Comment by dhiabenjaber** [comment]
![image](https://user-images.githubusercontent.com/74189546/168246626-51de1e87-498d-4635-a9b1-98b7e8c7d983.png)
this is the error that show up when i replace arrayAdapter with ApcuAdapter i work on a docker image and with php 7.2.34  i try to configure it but same problem not apcu not enabled

**Reply by oojacoboo** [reply]
You need to install it.

https://nsmith.net/apcu-with-docker-and-php-fpm

**Reply by oojacoboo** [reply]
Also, you should update PHP.  7.2 is out of date and won’t be supported soon.

**Comment by dhiabenjaber** [comment]
![image](https://user-images.githubusercontent.com/74189546/169019960-fbf48d3f-a466-46d2-8b1c-2a55e7ea736c.png)
i tried the apcu adapter i think that the time is less than the other time but it steel very long (1.5s) and i'm trying with graphql to speed up the response time compared with  the rest api .

**Comment by oojacoboo** [comment]
@dhiabenjaber This callstack analysis doesn't tell much of a story.  It looks like your autoloader is taking more time than needed though.  Are you using opcache?  The autoload should be cached.  With 7.4 you can pre-load as well (https://www.php.net/manual/en/opcache.preloading.php)

You need to show what's taking more time.  If 10% on autoload is your biggest issue... that's your biggest issue.  If it's something else... it's not in the screenshot.

**Comment by dhiabenjaber** [comment]
![image](https://user-images.githubusercontent.com/74189546/170736835-56d754b4-d7f3-446a-9e12-cd9ffd27d0dc.png)
@oojacoboo  as you can see here that the  function visit in graphql\language\visitor take a lot of time is there any solution

**Comment by oojacoboo** [comment]
@dhiabenjaber you can see here what that is:

https://github.com/webonyx/graphql-php/blob/master/src/Language/Visitor.php

This is part of the webonyx lib that GraphQLite is built upon.
